### PR TITLE
add ability to insert raw string in keymap.c

### PR DESCRIPTION
**Pull request by tpeacock19**
This adds the ability so insert a raw keycode string into a keymap. I
used this prior to adding the ability to insert a tap_dance_fn. I
figured that it may be useful for others.

### CODE REVIEW

**Comment by mihaiolteanu** [comment]
Hello @tpeacock19,

Thank you for your input!
I'm very glad you're enjoying mugur and found new use case for it.

To keep the project consistent, documented and its ease of discoverability by new potential users,
- Please consider updating the README as well, and give at least one example of how would one use this new key.
- Please also consider adding tests for this new entry (see the last section of mugur.el) and make sure to run all the projects tests before commiting.
- Please consider updating the docstring for `mugur--raw`. Currently it is,
```
  "Handle a modifier MUGUR-KEY.
...
```
which is the same as the docstring for `mugur--modifier`.

We'll take a look at your other PR after we succesfully close this one.
Happy typing!

**Comment by tpeacock19** [comment]
Ah right, of course. I did this in a bit of a rush and will update it soon. I'll have a few more coming as well. 

Also, you may want to split the tests into a separate file. Since there is no `(require 'ert)` it throws undefined errors when loading `mugur.el` if that library isn't already loaded.

File: README.md [hunk 1]
```diff
@@ -218,6 +218,14 @@ the mnemonic and the cadr is any valid mugur-key.
 ```
 
 ## Insert raw C code (advanced users)
+In case you have custom keycodes that you have created or simply want to
+insert a keycode that may be covered by mugur you can use the RAW mugur-key.
```
Comment (mihaiolteanu) [inline README.md:222]: may **not**?!

File: mugur.el [hunk 2]
```diff
@@ -1150,7 +1158,9 @@ that mugur-key."
   ((DF base)       "DF(BASE)")
   ((DF "base")     "DF(BASE)")
   ((LM "base" C)   "LM(BASE, MOD_LCTL)")
-  ((LT "base" x)   "LT(BASE, KC_X)")))
+  ((LT "base" x)   "LT(BASE, KC_X)")
+  ((RAW "custom_td_fn" "custom_td_fn"))
+  ((RAW "KC_ASUP")   "KC_ASUP")))
```
Comment (mihaiolteanu) [inline mugur.el:1163]: We can, of course, add KC_ASUP to `mugur--symbol`, right?

File: README.md [hunk 3]
```diff
@@ -218,6 +218,14 @@ the mnemonic and the cadr is any valid mugur-key.
...
+
+| mugur-key        | example comment                       |
+|:-----------------|:--------------------------------------|
+| (RAW "KC_MYKEY") | Insert your unmodified custom keycode |
```
Comment (mihaiolteanu) [inline README.md:226]: I'm still failing to see what a custom keycode would look like. Please give an example that you've personally used.

File: README.md (hunk 1 above)
Comment (tpeacock19) [reply README.md:222]: yeah i'll update this.

File: mugur.el (hunk 2 above)
Comment (tpeacock19) [reply mugur.el:1163]: We can add all the keycodes that are missing, yes. There's another PR that I'll eventually add with a few missing ones.

File: README.md (hunk 3 above)
Comment (tpeacock19) [reply README.md:226]: Personally, I've used RAW to refer a few missing keycodes, (DT_PRNT, DT_UP/DOWN, KC_ASRP/ASUP/ASDN) as well as functions in Tap Dances. However, I'm still working on trying to incorporate `ACTION_TAP_DANCE_LAYER_MOVE` and `ACTION_TAP_DANCE_LAYER_TOGGLE` so I haven't finished updating that PR quite yet.

But you can see more about custom keycodes here: https://github.com/qmk/qmk_firmware/blob/master/docs/custom_quantum_functions.md

**Comment by mihaiolteanu** [comment]
@tpeacock19 thank you for your input!
I do think the consistency and maturity of the code could be better, so I'll reject your PRs for the moment.
I greatly like your RAW keycode feature and idea. As the qmk expands (and some of the keycodes are missing from mugur) this is a must. So I'm looking forward for a single clean PR that addresses only this one issue, together with update readme and tests. This PR is close to it. Let me know when its ready for review.

Kind regards,
Mihai

**Comment by tpeacock19** [comment]
> I do think the consistency and maturity of the code could be better,
> so I'll reject your PRs for the moment.

By this do you mean the consistency in this specific PR or all of them?
Considering this PR simply has one pcase in it I have to imagine it is
the others. Which would lead me to ask what you mean by consistency and
maturity of the code?

Do you mean I should use expressions like these:

``` emacs-lisp
(defun mugur--qmk-keymap-layers (qmk-keymap)
  "Extract the qmk-layers from a QMK-KEYMAP."
  (aand (--find (eq (car it) 'layers)
                qmk-keymap)
        (cadr it)))
```

instead of something simple like this:

``` emacs-lisp
(defun mugur--qmk-keymap-layers (qmk-keymap)
  "Extract the qmk-layers from a QMK-KEYMAP."
  (car (alist-get 'layers qmk-keymap)))
```

> So I'm looking forward for a single clean PR that addresses only this
> one issue, together with update readme and tests. This PR is close to
> it.

Again, this PR only addresses that one issue of inserting raw strings
into the matrix in keymap.c. It includes two additions to your existing
test and an update to the readme.

However, I will close everything and look to go in another direction
since it does not seem you are interested in my contributions or working
together on them.

**Comment by mihaiolteanu** [comment]
> Again, this PR only addresses that one issue of inserting raw strings
> into the matrix in keymap.c. It includes two additions to your existing
> test and an update to the readme.

Unfortunately, you've removed the commits, but they've looked like this (see above)

```
 - ((LT "base" x)   "LT(BASE, KC_X)")))
 + ((LT "base" x)   "LT(BASE, KC_X)")
 + ((RAW "custom_td_fn" "custom_td_fn"))
 + ((RAW "KC_ASUP")   "KC_ASUP")))
```

As stated elsewhere, I prefer to stay on the safe side regarding code.
It is not the first "error" (I don't know if I should call it that) that I'm spotting on a first cursory view.

Again, thanks you for your contributions.
The raw keycode feature is really valuable. It's something that I haven't thought about before.

**Comment by tpeacock19** [comment]
This wouldn't result in any error because that keymap is not actually
being compiled in the test. Given that `mugur-valid-keycodes` is only
testing for keycode generation.

**Pull request closed without merging**